rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,307 @@
+    /// - There must be only one instance of `leased_lock::Client` per lease.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`LeaseNameLengthZero`](ErrorKind::LeaseNameLengthZero) if the `lease_name` is empty
+    ///
+    /// [`struct@Error`] of kind [`LeaseHolderNameLengthZero`](ErrorKind::LeaseHolderNameLengthZero) if the `lease_holder_name` is empty
+    pub fn new(
+        state_store: Arc<state_store::Client<C>>,
+        lease_name: Vec<u8>,
+        lease_holder_name: Vec<u8>,
+    ) -> Result<Self, Error> {
+        if lease_name.is_empty() {
+            return Err(Error(ErrorKind::LeaseNameLengthZero));
+        }
+
+        if lease_holder_name.is_empty() {
+            return Err(Error(ErrorKind::LeaseHolderNameLengthZero));
+        }
+
+        Ok(Self {
+            state_store,
+            lease_name,
+            lease_holder_name,
+            current_fencing_token: Arc::new(Mutex::new(None)),
+            auto_renewal_cancellation_token: CancellationToken::new(),
+        })
+    }
+
+    /// Gets the latest fencing token related to the most recent lease.
+    ///
+    /// Returns either None or an actual Fencing Token (`HybridLogicalClock`).
+    /// None means that either a lease has not been acquired previously with this client, or
+    /// a lease renewal has failed (if lease auto-renewal is used). The presence of a `HybridLogicalClock`
+    /// does not mean that it is the most recent (and thus valid) Fencing Token - this can
+    /// happen in the scenario where auto-renewal has not been used and the lease has already expired.
+    ///
+    /// # Panics
+    /// If the lock on the `current_fencing_token` is poisoned, which should not be possible.
+    #[must_use]
+    pub fn current_lease_fencing_token(&self) -> Option<HybridLogicalClock> {
+        self.current_fencing_token.lock().unwrap().clone()
+    }
+
+    async fn internal_acquire(
+        &self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+    ) -> Result<HybridLogicalClock, Error> {
+        let state_store_response = self
+            .state_store
+            .set(
+                self.lease_name.clone(),
+                self.lease_holder_name.clone(),
+                request_timeout,
+                None,
+                SetOptions {
+                    set_condition: SetCondition::OnlyIfEqualOrDoesNotExist,
+                    expires: Some(lease_expiration),
+                },
+            )
+            .await?;
+
+        if state_store_response.response {
+            self.current_fencing_token
+                .lock()
+                .unwrap()
+                .clone_from(&state_store_response.version);
+
+            state_store_response
+                .version
+                .ok_or(Error(ErrorKind::MissingFencingToken))
+        } else {
+            *self.current_fencing_token.lock().unwrap() = None;
+
+            Err(Error(ErrorKind::KeyAlreadyLeased))
```

Comment:
```suggestion
            Err(Error(ErrorKind::LeaseAlreadyHeld))
```
to keep away from the Key semantics?

Reply:
Done.